incubator/hnc/third_party/open-policy-agent/gatekeeper/pkg/webhook/certs.go

```diff
@@ -46,31 +46,37 @@ var crLog = logf.Log.WithName("cert-rotation")
 
 var vwhGVK = schema.GroupVersionKind{Group: "admissionregistration.k8s.io", Version: "v1beta1", Kind: "ValidatingWebhookConfiguration"}
 
+var crdGVK = schema.GroupVersionKind{Group: "apiextensions.k8s.io", Version: "v1beta1", Kind: "CustomResourceDefinition"}
+
 var _ manager.Runnable = &CertRotator{}
 
 var restartOnSecretRefresh = false
 
+var caInjectionAttempted = false
+
```

Comment:
Shouldn't this be part of the one of the objects rather than a global? E.g. if someone calls AddRotator twice, this value will be shared between the two calls, which is wrong.

Reply:
Also:
* Why not make this a chan like everything else, rather than spinlocking on line 635? (Spinlocks are usually a bad idea since they don't let other goroutines run).
* Is it really "Attempted?" I thought we only set this once it's *succeeded*?

Reply:
Ah right agreed to move it into object.

Answers to the other two questions:
1) I tried channel first and thought I could use it the same way as flipping this boolean here but found that I cannot close a closed channel and I will have to add a exponential backoff check like [this one](https://github.com/kubernetes-sigs/multi-tenancy/blob/hnc-v0.5/incubator/hnc/third_party/open-policy-agent/gatekeeper/pkg/webhook/certs.go#L526) on every WH. So I thought it might be easier to use a variable than channel.
2) It seems too complicated to check every single webhook considering we have 5 VWH and 3 CRD conversion WH. So I made it into `attempted` so that when the ca-injection reconciler finishes all `injectCertTo...` function calls, it's set to true.

WDYT?

Reply:
Re #2: I think if the call to `injectCertTo...` succeeds, then we _know_ it's worked, right? Why might there still be a problem if none of the injections returned an error?

Re #1: I see, for the other cases, it's really a one-time operation, but the reconciler can be called many times. Is that correct? If so, that makes sense but please comment that carefully and use an exponential backoff. Or maybe consider a channel with https://golang.org/pkg/sync/#Once.

Reply:
Discussed offline, 2) Yes, and since we can only flip it if no error is returned, I will rename it to `caInjected` instead of attempted; 1) I will replace the spinlocks with an exponential backoff to check when `caInjected` is flipped.

Reply:
Updated. Added `Tested by manually setting caInjected as false, the manager
pod kept restarting with error - cannot inject certs.` in the commit/PR message too.